### PR TITLE
Stateless execution prototype

### PR DESCRIPTION
**Pull request by s1na**
The goal of this PR is to prototype something akin to the [stateless client](https://ethresear.ch/t/the-stateless-client-concept/172) concept. There are two "roles":

- **Verifier** a light-client who wishes to process a future tx/block, but only has access to the current state root, and stores none of the actual state.
- **Prover** a full-node who runs a tx/block and generates all the necessary merkle proofs for the verifier to execute the tx. This includes proofs for accounts that were read during execution, codes for executed contracts and proofs for contract storage keys that were read.

The prover currently works by hooking the `StateManager` and generating proofs on invocation of `getAccount`, `getContractCode` and `getContractStorage`.

The verifier currently works by storing all received trie nodes in a new trie's db (with their hashes as keys), and using that trie in a new `StateManager`, and then running the tx.

To test this, I'm running the `GeneralStateTests` suite, running each case twice. Once to generate all the proofs, and a second time to test the verifier given proofs generated during the first run. There are still some edge cases that need to be fixed (38 cases are failing).

This also depends on a merkle-patricia-tree release (specifically https://github.com/ethereumjs/merkle-patricia-tree/pull/82, and probably https://github.com/ethereumjs/merkle-patricia-tree/pull/94).

### CODE REVIEW

**Comment by s1na** [comment]
It turns out my initial approach was overly complicated and prone to edge cases. I managed to bring down the initial ~600 failing test cases to 18, but then realized there's a better approach:

Instead of hooking `StateManager`'s methods (e.g. `getAccount`), we directly hook the underlying `leveldb.get`. The intuition is that every piece of state is stored in the db (e.g. accounts, contract codes, contract storage items) as a Buffer, with its `keccak256` hash as the key. So every time the `EVM` tries to fetch something from the db, if that item existed in the db before starting the transaction, necessarily the verifier would need that item. Items that did not exist before starting the transaction have been created during the tx, and the verifier could re-create them on its own.

The prover then collects all the pre-existing state items fetched from the db and sends them to verifier. The verifier saves all of these items in a new `leveldb` with their `keccak256` hash as key, and instantiates a Trie which uses this db and has the `preStateRoot` (correct state root prior to exeucting this tx) as root, and then runs the tx normally.

**Comment by s1na** [comment]
Currently all of the above is happening in the `StatelessRunner` test runner. I'm now wondering if we should add this feature to the library and if so what would be the best API. @holgerd77 @alcuadrado Any comments?

One option I can think of is to have a `ProofGeneratingVM` which does the hooking, and additionally returns all the proofs on `runTx` and `runBlock`; and a `StatelessVM` which should be instantiated with the proofs and then can do `runTx` and `runBlock` normally.

**Comment by holgerd77** [comment]
@s1na My understanding of the whole process structure is still limited, but my first impression is that it makes very much sense to add some experimental API on this and therefore integrate as a feature.

Would these two new `VM` classes subclasses of the `VM` class?

**Comment by s1na** [comment]
> Would these two new VM classes subclasses of the VM class?

Yeah probably. It's also possible to have a simpler API for now. For example a `proveTx(vm, txOpts)`, and `verifyStatelessTx(txOpts, proofs)`. Though I think sub-classing VM (at least for the verifying side) is more flexible and unlocks more use-cases.

**Comment by holgerd77** [comment]
Yeah, let's do the subclassing. This is really inviting for experimentation, and can't hurt if we clearly mark this as experimental.

**Comment by holgerd77** [comment]
Just to get a bit up-to-date: what's your plan on this? Do you want to get this out with `v4.1`?

**Comment by s1na** [comment]
I was testing the logic a bit more last week. I think I can get something ready for v4.1.

**Comment by holgerd77** [comment]
Rebased this, adopted this to the monorepo structure and updated the code to play well with changes within the test environment and directly use async functionality in several cases, so that test runs can now be executed again with `npm run test:stateless`.

Core logic itself needs still to be adopted.

**Comment by s1na** [comment]
@holgerd77 Thanks for updating it! I was just reading the thread and I'm thinking maybe as a first approach just merge this as a test runner so we can come up with a way of integrating it into the library? Potentially we could move the stateless logic to a package under the monorepo which depends on the VM.

In that case I'll have to look and see what's missing, if all tests pass etc. I don't remember what was the status of the PR last time I worked on it.

**Comment by holgerd77** [comment]
@s1na Thanks, I was assuming we would be following this `VM` subclassing approach as discussed initially, but open for other ways of structural integration as well. Think there were some last test cases not working yet.

File: packages/vm/tests/StatelessRunner.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    common: common
+  })
+  if (options.jsontrace) {
+    hookVM(vm, t)
+  }
+
+  // Determine set of all node hashes in the database
+  // before running the tx.
+  const existingKeys = new Set()
+  const it = stateManager._trie.db.iterator()
```
Comment (holgerd77) [inline packages/vm/tests/StatelessRunner.js:45]: Can you also give me a hint, where this `iterator()` might have moved to? This was errored as not available when I last ran the code, got stock there to some extend.

Comment (s1na) [reply packages/vm/tests/StatelessRunner.js:45]: This was possibly a local change, can't find an iterator method in MPT too. Levelup has an `iterator` [method](https://github.com/Level/levelup#dbiteratoroptions) which I probably used. Maybe just replacing that line with `stateManager._trie.db._leveldb.iterator()` would work.

File: packages/vm/tests/StatelessRunner.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  }
+
+  // Prepare tx and block
+  let tx = makeTx(testData.transaction)
+  let block = makeBlockFromEnv(testData.env)
+  tx._homestead = true
+  tx.enableHomestead = true
+  block.isHomestead = function () {
+    return true
+  }
```
Comment (ryanio) [inline packages/vm/tests/StatelessRunner.js:23]:
```suggestion
```

I don't think these homstead lines are needed, as I've been removing them in other refactors/PRs without issue.

**Comment by acolytec3** [comment]
Is there value in keeping this PR open now that we're looking towards verkle tries as the way forward on stateless execution?

**Comment by holgerd77** [comment]
Yeah, let's maybe keep at least *a bit* more open. I've the impression we are still extremely early with our own efforts and this might serve and some point of reference and/or inspiration. Once we are a bit further down the road and can judge better on the efforts made here we can close.
